src/template.c

```diff
@@ -5956,22 +5956,20 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
```

Comment:
If `inst` was done in speculative, it's already inserted in root module members. So the second time insertion is necessary only for non-root instances.